### PR TITLE
Support global attribute type

### PR DESCRIPTION
**Pull request by chrismccord**

### CODE REVIEW

File: lib/phoenix_component.ex [hunk 1]
```diff
@@ -284,16 +284,163 @@ defmodule Phoenix.Component do
   where the map contains all slot attributes, allowing us to access
   the label as `col.label`. This gives us complete control over how
   we render them.
+
+  ## Attributes
+  TODO
+
+  ### Global Attributes
```
Comment (josevalim) [inline lib/phoenix_component.ex:291]:
```suggestion
  ### Global Attributes

```

File: lib/phoenix_component.ex [hunk 2]
```diff
@@ -284,16 +284,163 @@ defmodule Phoenix.Component do
...
+
+  for prefix <- @global_prefixes do
+    def __global__?(unquote(prefix) <> _), do: true
+  end
+
+  def __global__?(_), do: false
+
+  @doc false
+  defmacro __using__(opts \\ []) do
     quote do
```
Comment (josevalim) [inline lib/phoenix_component.ex:420]: Use bind_quoted so you can get rid of the module attribute.

File: lib/phoenix_component.ex [hunk 3]
```diff
@@ -352,28 +499,26 @@ defmodule Phoenix.Component do
...
-      message = "attribute names must be atoms, got: #{inspect(name)}"
-      raise CompileError, line: line, file: file, description: message
+      compile_error!(line, file, "attribute names must be atoms, got: #{inspect(name)}")
     end
 
     unless is_list(opts) do
-      message = "expected attr/3 to receive keyword list of options, but got #{inspect(opts)}"
-      raise CompileError, line: line, file: file, description: message
+      compile_error!(line, file, """
+      expected attr/3 to receive keyword list of options, but got #{inspect(opts)}")
```
Comment (josevalim) [inline lib/phoenix_component.ex:507]:
```suggestion
      expected attr/3 to receive keyword list of options, but got #{inspect(opts)}"\
```

File: lib/phoenix_component.ex [hunk 4]
```diff
@@ -385,43 +530,62 @@ defmodule Phoenix.Component do
...
   @valid_types [:any] ++ @builtin_types
 
-  defp validate_attr_type!(name, type, line, file) when is_atom(type) do
+  defp validate_attr_type!(module, name, type, line, file) when is_atom(type) do
+    attrs = get_attrs(module)
+
+    cond do
+      Enum.find(attrs, fn attr -> attr.name == name end) ->
+        compile_error!(line, file, """
+        a duplicate attribute with name #{inspect(name)} already exists"
```
Comment (josevalim) [inline lib/phoenix_component.ex:542]:
```suggestion
        a duplicate attribute with name #{inspect(name)} already exists\
```

File: lib/phoenix_component.ex [hunk 5]
```diff
@@ -385,43 +530,62 @@ defmodule Phoenix.Component do
...
+        """)
+
+      existing = type == :global && Enum.find(attrs, fn attr -> attr.type === :global end) ->
```
Comment (josevalim) [inline lib/phoenix_component.ex:545]:
```suggestion
      existing = type == :global && Enum.find(attrs, fn attr -> attr.type == :global end) ->
```

File: lib/phoenix_component.ex [hunk 6]
```diff
@@ -284,16 +284,163 @@ defmodule Phoenix.Component do
...
+
+  Custom attribute prefixes can be provided by the caller module with
+  the `:globals` option to `use Phoenix.Component`. For example, the
+  following would allow Alpine JS annotations, such as `x-on:click`,
+  `x-data`, etc:
+
+      use Phoenix.Component, globals: ~w(x-)
   '''
 
+  @global_prefixes ~w(
```
Comment (josevalim) [inline lib/phoenix_component.ex:305]: Split prefixes vs static.

File: lib/phoenix_component.ex [hunk 7]
```diff
@@ -511,10 +683,28 @@ defmodule Phoenix.Component do
...
 
+        global_name = Enum.find_value(attrs, fn attr -> attr.type == :global && attr.name end)
+
+        known_keys =
+          for(attr <- attrs, do: attr.name) ++ Phoenix.LiveView.Helpers.__reserved_assigns__()
+
+        def_body =
+          if global_name do
+            quote do
+              {assgins, globals} = Map.split(assigns, unquote(known_keys))
```
Comment (josevalim) [inline lib/phoenix_component.ex:694]:
```suggestion
              {assigns, globals} = Map.split(assigns, unquote(known_keys))
```

File: lib/phoenix_component.ex [hunk 8]
```diff
@@ -385,43 +545,62 @@ defmodule Phoenix.Component do
...
   @valid_types [:any] ++ @builtin_types
 
-  defp validate_attr_type!(name, type, line, file) when is_atom(type) do
+  defp validate_attr_type!(module, name, type, line, file) when is_atom(type) do
+    attrs = get_attrs(module)
+
+    cond do
+      Enum.find(attrs, fn attr -> attr.name == name end) ->
+        compile_error!(line, file, """
+        a duplicate attribute with name #{inspect(name)} already exists\
```
Comment (feliperenan) [inline lib/phoenix_component.ex:557]:
```suggestion
        a duplicate attribute with name #{inspect(name)} already exists
```

I guess that's not intentional

File: lib/phoenix_component.ex [hunk 9]
```diff
@@ -511,10 +696,33 @@ defmodule Phoenix.Component do
...
+            nil -> {nil, nil}
+          end
+
+        known_keys =
+          for(attr <- attrs, do: attr.name) ++ Phoenix.LiveView.Helpers.__reserved_assigns__()
+
+        def_body =
+          if global_name do
+            quote do
+              {assgins, caller_globals} = Map.split(assigns, unquote(known_keys))
```
Comment (feliperenan) [inline lib/phoenix_component.ex:711]:
```suggestion
              {assigns, caller_globals} = Map.split(assigns, unquote(known_keys))
```

File: lib/phoenix_component.ex (hunk 8 above)
Comment (chrismccord) [reply lib/phoenix_component.ex:563]: José doesn't like unnecessary newlines, so intentional in this case :D

File: lib/phoenix_component.ex (hunk 9 above)
Comment (chrismccord) [reply lib/phoenix_component.ex]: 🙈

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**